models/issues/issue.go

```diff
@@ -121,6 +123,7 @@ type Issue struct {
 	NumComments      int
 	Ref              string
 	PinOrder         int `xorm:"DEFAULT 0"`
+	Version          int `xorm:"NOT NULL DEFAULT 0"`
```

Comment:
You can use an optimistic lock to do that.
https://xorm.io/docs/chapter-06/1.lock/

Reply:
done

Reply:
@lunny are we sure to use db level versioning. Is not it too risky to handle correct version info for all operations like api calls, pr operations, and issue updates:  can not we use version column for only content change?

```go
func UpdateIssueCols(ctx context.Context, issue *Issue, cols ...string) error {
	affected, err := db.GetEngine(ctx).ID(issue.ID).Cols(cols...).Update(issue)
	if err != nil {
		return err
	}
	if affected == 0 {
		return ErrIssueAlreadyChanged
	}
	issue.Version++
	return nil
}
```

```go
	case CommentTypeComment:
		if _, err = db.Exec(ctx, "UPDATE `issue` SET num_comments=num_comments+1 WHERE id=?", opts.Issue.ID); err != nil {
			return err
		}
		opts.Issue.Version++
		fallthrough
```

```go
func repoStatsCorrectIssueNumComments(ctx context.Context, id int64) error {
	return StatsCorrectSQL(ctx, "UPDATE `issue` SET num_comments=(SELECT COUNT(*) FROM `comment` WHERE issue_id=? AND type=0) WHERE id=?", id)
}
```

```go
	if _, err := sess.Exec("UPDATE `issue` SET `repo_id` = `repo_id` WHERE `id` = ?", pr.IssueID); err != nil {
		return false, err
	}
```

Reply:
We can use `NoVersionCheck()` on `UpdateIssueCols`.
```db.GetEngine(ctx).ID(issue.ID).Cols(cols...).NoVersionCheck().Update(issue)```

Only `Update` function will check the version. For `db.Exec`, it will not check the previous version.

Reply:
I do not think it's right to use XORM "version" here.

Otherwise, all issue-model code needs to call `NoVersionCheck`.

Actually, this PR's purpose is only to make "content" support version. And I think there are still many blocking problem, I will write a new comment.